### PR TITLE
[MCC-556901] Add specs for binary request bodies

### PR DESCRIPTION
**Pull request by cmcinnes-mdsol**
@mdsol/team-16 
@mdsol/architecture-enablement 

I realized our specs did not cover the authentication of requests with binary request bodies when I was pairing with Prajon on fixing [MCC-555926](https://jira.mdsol.com/browse/MCC-555926). I wanted to confirm that we fulfill this lin

> When hashing the request body, it should be read in as a byte array and not be converted in any way

from the spec [here](https://learn.mdsol.com/display/CA/MAuth%20Protocol%20V2%20Specification). 

The hashing functions in the ruby `Digest` module accept strings not byte arrays. Adding specs here using the `Digest::SHA512.file` method to read a binary file (a one pixel jpeg) into the hash function. Also confirmed that the SHA512 output is the same for this `blank.jpeg` file in C# as it is in Ruby with this code snippet I cobbled together (forgive any errors here this was my first brush with C#): 
```
using System;
using System.Security.Cryptography;
using System.Text;
using System.IO;

class MainClass {
  public static void Main (string[] args) {
    var jpgbytes = File.ReadAllBytes("blank.jpeg");
    var ba = SHA512.Create().ComputeHash(jpgbytes);
    Console.WriteLine (ByteArrayToString(ba));
  }

  public static string ByteArrayToString(byte[] ba)
  {
    StringBuilder hex = new StringBuilder(ba.Length * 2);
    foreach (byte b in ba)
      hex.AppendFormat("{0:x2}", b);
    return hex.ToString();
  }
}
```
The C# SHA512 hash function does take a byte array and matches up with the output of the Ruby hash function. This indicates that we do fulfill the spec.

Also removed `to_s` and replaced with a slightly more explicit `|| ''` for empty string handling as the request body and query string will only ever be `nil` or a string.

### CODE REVIEW

**Comment by ykitamura-mdsol** [comment]
I might not understand correctly 😅 but I don't get the point of adding these specs. Can you explain a little further?

> Also confirmed that the SHA512 output is the same for this blank.jpeg file in C# as it is in Ruby with this code snippet I cobbled together

How about adding specs for this against the `Signer#signature_v1` and `Signer#signature_v2` methods?

https://github.com/mdsol/mauth-client-ruby/blob/c40c8f4d83a9cb5e3a7930fff98e228614d02b18/lib/mauth/client/signer.rb#L55-L64

File: spec/client/local_authenticator_spec.rb [hunk 1]
```diff
@@ -102,6 +102,12 @@
           query_string: 'key=value&coolkey=coolvalue'
         )
       end
+      let(:binary_filepath) { 'spec/blank.jpeg' }
```
[the remark this answers is not in the document]
Comment (cmcinnes-mdsol) [reply spec/client/local_authenticator_spec.rb]: Sounds good, wasn't sure where to put it.

File: lib/mauth/request_and_response.rb [hunk 2]
```diff
@@ -59,9 +59,9 @@ def string_to_sign_v2(override_attrs)
 
       # memoization of body_digest to avoid hashing three times when we call
       # string_to_sign_v2 three times in client#signature_valid_v2!
-      # note that if :body is nil we hash an empty string ("")
-      attrs_with_overrides[:body_digest] ||= Digest::SHA512.hexdigest(attrs_with_overrides[:body].to_s)
-      attrs_with_overrides[:encoded_query_params] = encode_query_string(attrs_with_overrides[:query_string].to_s)
+      # note that if :body is nil we hash an empty string ('')
+      attrs_with_overrides[:body_digest] ||= Digest::SHA512.hexdigest(attrs_with_overrides[:body] || '')
+      attrs_with_overrides[:encoded_query_params] = encode_query_string(attrs_with_overrides[:query_string] || '')
```
[the remark this answers is not in the document]
Comment (cmcinnes-mdsol) [reply lib/mauth/request_and_response.rb]: Yep it's to be more obvious.

File: spec/client/local_authenticator_spec.rb [hunk 3]
```diff
@@ -214,6 +220,33 @@
...
+          request_url: '/',
+          body: binary_file_body,
+          query_string: 'key=value&coolkey=coolvalue'
+        )
+
+        # the signing mauth client should be able to stream large request bodies
+        # from the disk straight into the hashing funtion like so:
+        streamed_hash_digest = Digest::SHA512.file(binary_filepath).hexdigest
+        # used the digest from streaming in the file when signing the request
+        signed_request = client.signed(request, request_body_digest: streamed_hash_digest)
```
[the remark this answers is not in the document]
Comment (cmcinnes-mdsol) [reply spec/client/local_authenticator_spec.rb]: @ykitamura-mdsol It overrides the `body_digest` that would normally be generated in `string_to_sign_V2`. (Realized this should just be `:body_digest` and updated). In this case it allows us to simulate signing by a client that read the request body from disk.

[the remark this answers is not in the document]
Comment (ykitamura-mdsol) [reply spec/client/local_authenticator_spec.rb]: @cmcinnes-mdsol I see. How about adding a fail case to make sure mauth-client is using `:body_digest`, something like this?
```ruby
wrong_hash_digest = Digest::SHA512.hexdigest('abc')
signed_request = client.signed(request, body_digest: wrong_hash_digest)
expect(authenticating_mc.authentic?(signed_request)).to be false
```

[the remark this answers is not in the document]
Comment (cmcinnes-mdsol) [reply spec/client/local_authenticator_spec.rb]: Sure, added @ykitamura-mdsol

File: spec/client/local_authenticator_spec.rb [hunk 4]
```diff
@@ -214,6 +216,39 @@
         expect(authenticating_mc.authentic?(signed_request)).to be true
       end
 
+      it 'considers a signed request with a request body of binary data to be authentic' do
+        request = TestSignableRequest.new(
+          verb: 'PUT',
+          request_url: '/',
+          body: binary_file_body,
+          query_string: 'key=value&coolkey=coolvalue'
+        )
```
Comment (ykitamura-mdsol) [inline spec/client/local_authenticator_spec.rb]: How about DRY this up and use it in new specs?
```ruby
let(:binary_request) do
  TestSignableRequest.new(
    verb: 'PUT',
    request_url: '/',
    body: binary_file_body,
    query_string: 'key=value&coolkey=coolvalue'
  )
end
```

**Pull request merged**